### PR TITLE
test: fix testsuite [WPB-10397]

### PR DESCRIPTION
**Pull request by SimonThormeyer**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

_Briefly describe the issue you have solved or implemented with this pull request. If the PR contains multiple issues, use a bullet list._

### Causes (Optional)

_Briefly describe the causes behind the issues. This could be helpful to understand the adopted solutions behind some nasty bugs or complex issues._

### Solutions

_Briefly describe the solutions you have implemented for the issues explained above._

### Dependencies (Optional)

_If there are some other pull requests related to this one (e.g. new releases of frameworks), specify them here._

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,6 +16,25 @@ No license is granted to the Wire trademark and its associated logos, all of whi
...
+The integration tests are very flaky, mainly because the behavior of the Keycloak docker container is very flaky.
+This is why we need the `no-fail-fast` flag in the test command.
+This will have in the average case about 75 % of the tests pass.
+A possible approach to deal with the flakiness is to run the command below and then run the tests that failed in the
+first run individually.
+
+```bash
+cargo nextest run --no-fail-fast
+```
+
```
Comment (istankovic) [inline README.md:37]: Good to see more docs, but could you please factor out the README changes into their own commit?

File: e2e-identity/Cargo.toml [hunk 2]
```diff
@@ -10,6 +10,9 @@ publish = false
 [lib]
 crate-type = ["cdylib", "rlib"]
 
+[lints.rust]
+unexpected_cfgs = { level = "allow", check-cfg = ['cfg(ci)'] }
+
```
Comment (istankovic) [inline e2e-identity/Cargo.toml:15]: I think this should rather be addressed in `e2e-identity/build.rs`, where we tell rustc about `ci`.
See https://doc.rust-lang.org/cargo/reference/build-scripts.html#rustc-check-cfg.

File: e2e-identity/tests/utils/docker/dex.rs [hunk 3]
```diff
@@ -47,36 +49,36 @@ impl DexImage {
...
+    fn name(&self) -> &str {
+        Self::NAME
     }
 
-    fn tag(&self) -> String {
-        std::env::var("DEX_VERSION").unwrap_or_else(|_| Self::TAG.to_string())
+    fn tag(&self) -> &str {
+        std::env::var("DEX_VERSION")
+            .map(|tag| Box::leak(tag.into_boxed_str()) as &str)
+            .unwrap_or(Self::TAG)
```
Comment (istankovic) [inline e2e-identity/tests/utils/docker/dex.rs:68]: Why leak here? Why don't we get the tag at `DexImage` init time and then just give out a proper reference?

File: e2e-identity/tests/utils/docker/keycloak.rs [hunk 4]
```diff
@@ -266,42 +273,39 @@ impl From<&KeycloakCfg> for UserRepresentation {
...
+    fn name(&self) -> &str {
+        Self::NAME
     }
 
-    fn tag(&self) -> String {
-        env::var("KEYCLOAK_VERSION").unwrap_or_else(|_| Self::TAG.to_string())
+    fn tag(&self) -> &str {
+        env::var("KEYCLOAK_VERSION")
+            .map(|tag| Box::leak(tag.into_boxed_str()) as &str)
+            .unwrap_or_else(|_| Self::TAG)
```
Comment (istankovic) [inline e2e-identity/tests/utils/docker/keycloak.rs:283]: Similar comment as for `DexImage`.

File: e2e-identity/tests/utils/docker/ldap.rs [hunk 5]
```diff
@@ -55,47 +58,42 @@ impl LdapImage {
...
         Box::new(self.env_vars.iter())
     }
 
-    fn volumes(&self) -> Box<dyn Iterator<Item = (&String, &String)> + '_> {
-        Box::new(self.volumes.iter())
+    fn mounts(&self) -> impl IntoIterator<Item = &Mount> {
+        &self.volumes
+    }
+
+    fn cmd(&self) -> impl IntoIterator<Item = impl Into<Cow<'_, str>>> {
```
Comment (istankovic) [inline e2e-identity/tests/utils/docker/ldap.rs:84]: Shouldn't this be `impl Iterator`?? Given that you call `into_iter` below?

File: e2e-identity/tests/e2e.rs [hunk 6]
```diff
@@ -111,16 +106,16 @@ mod alg {
 
     #[tokio::test]
     async fn ed25519_should_succeed() {
-        let test = E2eTest::new_internal(false, JwsAlgorithm::Ed25519, OidcProvider::Dex)
-            .start(docker())
+        let test = E2eTest::new_internal(false, JwsAlgorithm::Ed25519, OidcProvider::Keycloak)
+            .start()
```
Comment (istankovic) [inline e2e-identity/tests/e2e.rs:110]: What's the reason we're switching to Keycloak here?

**Comment by istankovic** [comment]
Commit `bbd66686ac76faa1e1369a0b8907f19d90c6b850` says
```
test: upgrade tests to new testcontainers api 
```
while the one before it, `5706aff84c2b3d246642ad3077fea7448e1a207a` says
```
 test: update testcontainers crate 
```
but it contains not just changes to `Cargo.{toml,lock}`, but also a bunch of changes to tests related to the new testcontainers API. So I think the "update testcontainers crate" should only have changes to Cargo files, while all the test-related changes should go to the "upgrade tests to new testcontainers api" commit.

**Comment by istankovic** [comment]
The message of commit `99b70be21a708c23b906d1416e598c4d39021ec1` says
```
 test: improve behavior of stepca [WPB-10397]
```
but it doesn't really tell why is the behaviour improved, and I couldn't really understand it from the code diff, though it is small.

**Review by istankovic — CHANGES_REQUESTED** [review]
In general looks good, but there are few minor things to sort out.

File: e2e-identity/tests/utils/docker/ldap.rs (hunk 5 above)
Comment (SimonThormeyer) [reply e2e-identity/tests/utils/docker/ldap.rs:85]: This overrides a default trait implementation, so the whole function signature must match

File: e2e-identity/tests/utils/docker/ldap.rs [hunk 7]
```diff
@@ -55,47 +61,42 @@ impl LdapImage {
...
+    fn name(&self) -> &str {
+        Self::NAME
     }
 
-    fn tag(&self) -> String {
-        std::env::var("OPENLDAP_VERSION").unwrap_or_else(|_| Self::TAG.to_string())
+impl ImageArgs for LdapArgs {
+    fn into_iterator(self) -> Box<dyn Iterator<Item = String>> {
+    fn tag(&self) -> &str {
+        &self.tag
```
Comment (istankovic) [inline e2e-identity/tests/utils/docker/ldap.rs:70]: This is a very strange hunk. :-)

Comment (SimonThormeyer) [reply e2e-identity/tests/utils/docker/ldap.rs:70]: Let me factor out these changes into more commits, this is too big for a single one

File: e2e-identity/tests/utils/docker/wiremock.rs [hunk 8]
```diff
@@ -46,26 +44,28 @@ impl WiremockImage {
...
     }
 
-    fn volumes(&self) -> Box<dyn Iterator<Item = (&String, &String)> + '_> {
-        Box::new(self.volumes.iter())
+    fn mounts(&self) -> impl IntoIterator<Item = &Mount> {
+        &self.volumes
+    }
+
+    fn cmd(&self) -> impl IntoIterator<Item = impl Into<Cow<'_, str>>> {
+        vec!["/stubs".to_string(), "-p".to_string(), WiremockImage::PORT.to_string()].into_iter()
```
Comment (istankovic) [inline e2e-identity/tests/utils/docker/wiremock.rs:64]: Two questions here:
- what's `/stubs` and what does `/stubs -p` do?
- why call `into_iter` when `Vec` impls `IntoIterator`?

File: e2e-identity/tests/utils/docker/ldap.rs [hunk 9]
```diff
@@ -55,47 +61,42 @@ impl LdapImage {
...
+    }
+
+    fn cmd(&self) -> impl IntoIterator<Item = impl Into<Cow<'_, str>>> {
+        Box::new(
+            vec![
+                "--copy-service".to_string(),
+                "--loglevel".to_string(),
+                "debug".to_string(),
+            ]
+            .into_iter(),
```
Comment (istankovic) [inline e2e-identity/tests/utils/docker/ldap.rs:92]: I think this `into_iter` can be omitted, too.

File: e2e-identity/tests/e2e.rs (hunk 6 above)
Comment (istankovic) [reply e2e-identity/tests/e2e.rs:110]: Thanks for taking the time to pull this into a separate commit and writing a nice commit message!

**Review by istankovic — APPROVED** [review]
Great work!
